### PR TITLE
Prepend refs/heads/ to issue template refs

### LINKED ISSUE #20456: Issue Template Uses Incorrect Path For Ref (Commit Instead Of Branch)
### Description

When adding an issue from a template, the icon that links to the ref incorrectly wants to reference a commit instead of the branch, despite the branch existing with that same name. See the attached screenshot.

The template.md file contains:

```MD
---

name: "New Bug"
about: "Log a new bug!"
title: "Bug: "
ref: "main"
labels:

- bug

---
```

Expected path: `/user/repo/src/branch/main`
Actual path: `/user/repo/src/commit/main`

The actual path, which points to `commit` does not exist and thus results in a 404.

My understanding is that the link as per the screenshot will take you to the branch/tag as if you manually selected it when creating an issue and picking a branch/tag and the "ref" portion of the template would do exactly that. The documentation on issue templates was vague and doesn't mention what "ref" actually refers to.

### Screenshots

Repo_Issue_5

### Gitea Version

1.17.0-rc2

### Can you reproduce the bug on the Gitea demo site?

No

### Operating System

Windows 10

### Browser Version

Firefox 102

### PR DESCRIPTION
**Pull request by zeripath**
Fix #20456

At some point during the 1.17 cycle abbreviated refishs to issue branches started breaking. This is likely due serious inconsistencies in our management of refs throughout Gitea - which is a bug needing to be addressed in a different PR. (Likely more than one)

We should try to use non-abbreviated `fullref`s as much as possible. That is where a user has inputted a abbreviated `refish` we should add `refs/heads/` if it is `branch` etc. I know people keep writing and merging PRs that remove prefixes from stored content but it is just wrong and it keeps causing problems like this. We should only remove the prefix at the time of
presentation as the prefix is the only way of knowing umambiguously and permanently if the `ref` is referring to a `branch`, `tag` or `commit` / `SHA`. We need to make it so that every ref has the appropriate prefix, and probably also need to come up with some definitely unambiguous way of storing `SHA`s if they're used in a `ref` or `refish` field. We must not store a potentially
ambiguous `refish` as a `ref`. (Especially when referring a `tag` - there is no reason why users cannot create a `branch` with the same short name as a `tag` and vice versa and any attempt to prevent this will fail. You can even create a `branch` and a
`tag` that matches the `SHA` pattern.)

To that end in order to fix this bug, when parsing issue templates check the provided `Ref` (here a `refish` because almost all users do not know or understand the subtly), if it does not start with `refs/` add the `BranchPrefix` to it. This allows people to make their templates refer to a `tag` but not to a `SHA` directly. (I don't think that is particularly unreasonable but if people disagree I can make the `refish` be checked to see if it matches the `SHA` pattern.)

Next we need to handle the issue links that are already written. The links here are created with `git.RefURL`

Here we see there is a bug introduced in #17551 whereby the provided `ref` argument can be double-escaped so we remove the incorrect external escape. (The escape added in #17551 is in the right place - unfortunately I missed that the calling function was doing the wrong thing.)

Then within `RefURL()` we check if an unprefixed `ref` (therefore potentially a `refish`) matches the `SHA` pattern before assuming that is actually a `commit` - otherwise is assumed to be a `branch`. This will handle most of the problem cases excepting the very unusual cases where someone has deliberately written a `branch` to look like a `SHA1`.

But please if something is called a `ref` or interpreted as a `ref` make it a full-ref before storing or using it. By all means if something is a `branch` assume the prefix is removed but always add it back in if you are using it as a `ref`. Stop storing abbreviated `branch` names and `tag` names - which are `refish` as a `ref`. It will keep on causing problems like this.

Fix #20456

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by silverwind** [comment]
Is "refish" equal to `commit-ish` or `tree-ish` in git terminology? If so, I'd prefer that term :)

**Review by silverwind — APPROVED** [review]

**Comment by zeripath** [comment]
> Is "refish" equal to `commit-ish` or `tree-ish` in git terminology? If so, I'd prefer that term :)

I'm using it here as a catch all term for all of those things. But I am particularly trying to draw notice to the idea that `main` isn't a complete ref even if git would consider it a `<ref>`, `<commit-ish>`, and `<tree-ish>`.

**Comment by zeripath** [comment]
conflicts resolved

File: routers/web/repo/issue.go [hunk 1]
```diff
@@ -784,6 +784,10 @@ func setTemplateIfExists(ctx *context.Context, ctxDataKey string, possibleFiles
 					}
 				}
 			}
+
+		}
+		if !strings.HasPrefix(template.Ref, "refs/") { // Assume that the ref intended is always a branch - for tags users should use refs/tags/<ref>
```
Comment (delvh) [inline routers/web/repo/issue.go:789]: Why not
```suggestion
		if !strings.HasPrefix(template.Ref, git.BranchPrefix) { // Assume that the ref intended is always a branch - for tags users should use refs/tags/<ref>
```
?

File: modules/context/repo.go [hunk 2]
```diff
@@ -1087,6 +1087,9 @@ func (ctx *Context) IssueTemplatesErrorsFromDefaultBranch() ([]*api.IssueTemplat
 			if it, err := template.UnmarshalFromEntry(entry, dirName); err != nil {
 				invalidFiles[fullName] = err
 			} else {
+				if !strings.HasPrefix(it.Ref, "refs/") { // Assume that the ref intended is always a branch - for tags users should use refs/tags/<ref>
```
Comment (delvh) [inline modules/context/repo.go:1092]: Why not
```suggestion
				if !strings.HasPrefix(it.Ref, git.BranchPrefix) { // Assume that the ref intended is always a branch - for tags users should use refs/tags/<ref>
```
?

Comment (lunny) [reply modules/context/repo.go:1092]: It's not the same. `refs/heads` means branch not `refs`.

File: routers/web/repo/issue.go (hunk 1 above)
Comment (lunny) [reply routers/web/repo/issue.go:789]: As above.

File: modules/context/repo.go (hunk 2 above)
Comment (delvh) [reply modules/context/repo.go:1092]: yes. But have a look one line below the suggestion to see why I'm somewhat against the current implementation.

Comment (delvh) [reply modules/context/repo.go:1092]: Okay, yeah, I can see now that it won't work when trying to use tags.
Perhaps we should add a new Prefix: `const git.RefPrefix = "refs/"` and adapt the other prefixes to `RefPrefix + rest`

File: routers/web/repo/issue.go (hunk 1 above)
Comment (zeripath) [reply routers/web/repo/issue.go:789]: Because that won't work for tags. Read the associated comment.

File: modules/context/repo.go (hunk 2 above)
Comment (zeripath) [reply modules/context/repo.go:1092]: "refs/" is used in these two places only.

It seems OTT to create a const and frankly more confusing for that. 

The correct solution if we want to be strict is to create a series of types that would ensure that this was done correctly however that is a bigger job than this PR, (which I would note has itself already taken 3 months to get review.)

**Review by lafriks — APPROVED** [review]

File: modules/context/repo.go (hunk 2 above)
Comment (lunny) [reply modules/context/repo.go:1092]: @delvh Is this resolved?

Comment (delvh) [reply modules/context/repo.go:1092]: Yes

**Review by delvh — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**